{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MDu46olS8MyK",
        "outputId": "05851ddd-370c-407e-90d5-13e2ea81c566"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"GEOquery\", ask = FALSE)\n",
        "BiocManager::install(\"limma\", ask = FALSE)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ts2tYyAL9alR",
        "outputId": "08b1da47-f82a-4805-8dd9-0666abf86470"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.21 (BiocManager 1.30.25), R 4.5.0 (2025-04-11)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'GEOquery'\n",
            "\n",
            "also installing the dependencies ‘matrixStats’, ‘XVector’, ‘UCSC.utils’, ‘GenomeInfoDbData’, ‘abind’, ‘SparseArray’, ‘BiocGenerics’, ‘statmod’, ‘XML’, ‘R.oo’, ‘R.methodsS3’, ‘MatrixGenerics’, ‘GenomicRanges’, ‘IRanges’, ‘GenomeInfoDb’, ‘S4Arrays’, ‘DelayedArray’, ‘Biobase’, ‘limma’, ‘rentrez’, ‘R.utils’, ‘SummarizedExperiment’, ‘S4Vectors’\n",
            "\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.21 (BiocManager 1.30.25), R 4.5.0 (2025-04-11)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'limma'”\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "library(GEOquery)\n",
        "library(limma)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kqLtxh5KCLrK",
        "outputId": "5e49c9bd-22cf-40b4-a137-330b80e76af4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Loading required package: Biobase\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "Loading required package: generics\n",
            "\n",
            "\n",
            "Attaching package: ‘generics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
            "    setequal, union\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
            "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
            "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
            "    unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "Setting options('download.file.method.GEOquery'='auto')\n",
            "\n",
            "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
            "\n",
            "\n",
            "Attaching package: ‘limma’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:BiocGenerics’:\n",
            "\n",
            "    plotMA\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(\"umap\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5NGT3weyCkmt",
        "outputId": "532b3979-5088-4dac-8fd9-ab07fcae42d0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependencies ‘RcppTOML’, ‘here’, ‘png’, ‘RcppEigen’, ‘reticulate’, ‘RSpectra’\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "library(umap)\n"
      ],
      "metadata": {
        "id": "L8AAZp1qCmbv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Version info: R 4.2.2, Biobase 2.58.0, GEOquery 2.66.0, limma 3.54.0\n",
        "################################################################\n",
        "#   Differential expression analysis with limma\n",
        "library(GEOquery)\n",
        "library(limma)\n",
        "library(umap)\n",
        "\n",
        "# load series and platform data from GEO\n",
        "\n",
        "gset <- getGEO(\"GSE45827\", GSEMatrix =TRUE, AnnotGPL=TRUE)\n",
        "if (length(gset) > 1) idx <- grep(\"GPL570\", attr(gset, \"names\")) else idx <- 1\n",
        "gset <- gset[[idx]]\n",
        "\n",
        "# make proper column names to match toptable\n",
        "fvarLabels(gset) <- make.names(fvarLabels(gset))\n",
        "\n",
        "# group membership for all samples\n",
        "gsms <- paste0(\"11111111111111111111111111111111111111111111111XXX\",\n",
        "        \"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX0000000000000XXXX\",\n",
        "        \"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\",\n",
        "        \"XXXXX\")\n",
        "sml <- strsplit(gsms, split=\"\")[[1]]\n",
        "\n",
        "# filter out excluded samples (marked as \"X\")\n",
        "sel <- which(sml != \"X\")\n",
        "sml <- sml[sel]\n",
        "gset <- gset[ ,sel]\n",
        "\n",
        "# log2 transformation\n",
        "ex <- exprs(gset)\n",
        "qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))\n",
        "LogC <- (qx[5] > 100) ||\n",
        "          (qx[6]-qx[1] > 50 && qx[2] > 0)\n",
        "if (LogC) { ex[which(ex <= 0)] <- NaN\n",
        "  exprs(gset) <- log2(ex) }\n",
        "\n",
        "# assign samples to groups and set up design matrix\n",
        "gs <- factor(sml)\n",
        "groups <- make.names(c(\"Normal Controls\",\"Tumour Controls\"))\n",
        "levels(gs) <- groups\n",
        "gset$group <- gs\n",
        "design <- model.matrix(~group + 0, gset)\n",
        "colnames(design) <- levels(gs)\n",
        "\n",
        "gset <- gset[complete.cases(exprs(gset)), ] # skip missing values\n",
        "\n",
        "fit <- lmFit(gset, design)  # fit linear model\n",
        "\n",
        "# set up contrasts of interest and recalculate model coefficients\n",
        "cts <- paste(groups[1], groups[2], sep=\"-\")\n",
        "cont.matrix <- makeContrasts(contrasts=cts, levels=design)\n",
        "fit2 <- contrasts.fit(fit, cont.matrix)\n",
        "\n",
        "# compute statistics and table of top significant genes\n",
        "fit2 <- eBayes(fit2, 0.01)\n",
        "tT <- topTable(fit2, adjust=\"fdr\", sort.by=\"B\", number=250)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gI1-AS7TFxSO",
        "outputId": "8dbbf8f2-5048-46c9-99bf-45cf4e929975"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Found 1 file(s)\n",
            "\n",
            "GSE45827_series_matrix.txt.gz\n",
            "\n",
            "Using locally cached version: /tmp/RtmplgehwL/GSE45827_series_matrix.txt.gz\n",
            "\n",
            "Using locally cached version of GPL570 found here:\n",
            "/tmp/RtmplgehwL/GPL570.annot.gz \n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Differential expression analysis with limma\n",
        "library(GEOquery)\n",
        "library(limma)\n",
        "library(umap)\n",
        "\n",
        "# Load series and platform data from GEO\n",
        "gset <- getGEO(\"GSE45827\", GSEMatrix = TRUE, AnnotGPL = TRUE)\n",
        "if (length(gset) > 1) idx <- grep(\"GPL570\", attr(gset, \"names\")) else idx <- 1\n",
        "gset <- gset[[idx]]\n",
        "\n",
        "# Make proper column names\n",
        "fvarLabels(gset) <- make.names(fvarLabels(gset))\n",
        "\n",
        "# Group membership for all samples\n",
        "gsms <- paste0(\"11111111111111111111111111111111111111111111111XXX\",\n",
        "               \"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX0000000000000XXXX\",\n",
        "               \"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\",\n",
        "               \"XXXXX\")\n",
        "sml <- strsplit(gsms, split = \"\")[[1]]\n",
        "\n",
        "# Filter out excluded samples (marked as \"X\")\n",
        "sel <- which(sml != \"X\")\n",
        "sml <- sml[sel]\n",
        "gset <- gset[, sel]\n",
        "\n",
        "# Log2 transformation if necessary\n",
        "ex <- exprs(gset)\n",
        "qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm = TRUE))\n",
        "LogC <- (qx[5] > 100) || (qx[6] - qx[1] > 50 && qx[2] > 0)\n",
        "if (LogC) {\n",
        "  ex[which(ex <= 0)] <- NaN\n",
        "  exprs(gset) <- log2(ex)\n",
        "}\n",
        "\n",
        "# Assign samples to groups and set up design matrix\n",
        "gs <- factor(sml)\n",
        "groups <- make.names(c(\"Normal Controls\", \"Tumour Controls\"))\n",
        "levels(gs) <- groups\n",
        "gset$group <- gs\n",
        "design <- model.matrix(~ group + 0, gset)\n",
        "colnames(design) <- levels(gs)\n",
        "\n",
        "# Remove missing values\n",
        "gset <- gset[complete.cases(exprs(gset)), ]\n",
        "\n",
        "# Fit linear model\n",
        "fit <- lmFit(gset, design)\n",
        "\n",
        "# Set up contrasts\n",
        "cts <- paste(groups[1], groups[2], sep = \"-\")\n",
        "cont.matrix <- makeContrasts(contrasts = cts, levels = design)\n",
        "fit2 <- contrasts.fit(fit, cont.matrix)\n",
        "fit2 <- eBayes(fit2, 0.01)\n",
        "\n",
        "# Top table\n",
        "tT <- topTable(fit2, adjust = \"fdr\", sort.by = \"B\", number = 250)\n",
        "tT <- tT$ID <- rownames(tT)\n",
        "annotated_tT <- merge(tT, fData(gset), by = \"ID\")\n",
        "subset(annotated_tT, select = c(\"ID\", \"logFC\", \"P.Value\", \"adj.P.Val\", \"Gene.symbol\", \"Gene.title\"))\n",
        "\n",
        "write.table(tT, file = stdout(), row.names = FALSE, sep = \"\\t\")\n",
        "\n",
        "# Histogram of adjusted P-values\n",
        "tT2 <- topTable(fit2, adjust = \"fdr\", sort.by = \"B\", number = Inf)\n",
        "hist(tT2$adj.P.Val, col = \"grey\", border = \"white\",\n",
        "     xlab = \"Adjusted P-value\", ylab = \"Number of genes\",\n",
        "     main = \"P-adj value distribution\")\n",
        "\n",
        "# Decide test results\n",
        "dT <- decideTests(fit2, adjust.method = \"fdr\", p.value = 0.05, lfc = 0)\n",
        "\n",
        "# Venn diagram\n",
        "vennDiagram(dT, circle.col = palette())\n",
        "\n",
        "# Q-Q plot\n",
        "t.good <- which(!is.na(fit2$F))\n",
        "qqt(fit2$t[t.good], fit2$df.total[t.good], main = \"Moderated t statistic\")\n",
        "\n",
        "# Volcano plot\n",
        "colnames(fit2)\n",
        "ct <- 1\n",
        "volcanoplot(fit2, coef = ct, main = colnames(fit2)[ct], pch = 20,\n",
        "            highlight = length(which(dT[, ct] != 0)), names = rep('+', nrow(fit2)))\n",
        "\n",
        "# MD plot\n",
        "plotMD(fit2, column = ct, status = dT[, ct], legend = FALSE, pch = 20, cex = 1)\n",
        "abline(h = 0)\n",
        "\n",
        "# General expression plots\n",
        "ex <- exprs(gset)\n",
        "\n",
        "# Boxplot\n",
        "dev.new(width = 3 + ncol(gset) / 6, height = 5)\n",
        "ord <- order(gs)\n",
        "palette(c(\"#1B9E77\", \"#7570B3\", \"#E7298A\", \"#E6AB02\", \"#D95F02\",\n",
        "          \"#66A61E\", \"#A6761D\", \"#B32424\", \"#B324B3\", \"#666666\"))\n",
        "par(mar = c(7, 4, 2, 1))\n",
        "boxplot(ex[, ord], boxwex = 0.6, notch = TRUE,\n",
        "        main = paste(\"GSE45827 /\", annotation(gset)),\n",
        "        outline = FALSE, las = 2, col = gs[ord])\n",
        "legend(\"topleft\", groups, fill = palette(), bty = \"n\")\n",
        "dev.off()\n",
        "\n",
        "# Density plot\n",
        "par(mar = c(4, 4, 2, 1))\n",
        "plotDensities(ex, group = gs, main = paste(\"GSE45827 /\", annotation(gset), \"value distribution\"), legend = \"topright\")\n",
        "\n",
        "# UMAP plot\n",
        "ex <- na.omit(ex)\n",
        "ex <- ex[!duplicated(ex), ]\n",
        "ump <- umap(t(ex), n_neighbors = 15, random_state = 123)\n",
        "par(mar = c(3, 3, 2, 6), xpd = TRUE)\n",
        "plot(ump$layout, main = \"UMAP plot, nbrs=15\", xlab = \"\", ylab = \"\", col = gs, pch = 20, cex = 1.5)\n",
        "legend(\"topright\", inset = c(-0.15, 0), legend = levels(gs), pch = 20,\n",
        "       col = 1:nlevels(gs), title = \"Group\", pt.cex = 1.5)\n",
        "\n",
        "# Mean-variance trend plot\n",
        "plotSA(fit2, main = \"Mean variance trend, GSE45827\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 506
        },
        "id": "XPA97aIlGwwG",
        "outputId": "c3bd0f65-ba44-4d89-d327-4275a48b328b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Found 1 file(s)\n",
            "\n",
            "GSE45827_series_matrix.txt.gz\n",
            "\n",
            "Using locally cached version: /tmp/RtmplgehwL/GSE45827_series_matrix.txt.gz\n",
            "\n",
            "Using locally cached version of GPL570 found here:\n",
            "/tmp/RtmplgehwL/GPL570.annot.gz \n",
            "\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "Error in fix.by(by.x, x): 'by' must specify a uniquely valid column\n",
          "traceback": [
            "Error in fix.by(by.x, x): 'by' must specify a uniquely valid column\nTraceback:\n",
            "1. merge.default(tT, fData(gset), by = \"ID\")",
            "2. merge(as.data.frame(x), as.data.frame(y), ...)",
            "3. merge.data.frame(as.data.frame(x), as.data.frame(y), ...)",
            "4. fix.by(by.x, x)",
            "5. stop(ngettext(sum(bad), \"'by' must specify a uniquely valid column\", \n .     \"'by' must specify uniquely valid columns\"), domain = NA)",
            "6. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"'by' must specify a uniquely valid column\", base::quote(fix.by(by.x, \n .     x)))"
          ]
        }
      ]
    }
  ]
}